rustls/src/conn.rs

```diff
@@ -358,6 +358,26 @@ https://docs.rs/rustls/latest/rustls/manual/_03_howto/index.html#unexpected-eof"
 #[cfg(feature = "std")]
 pub use connection::{Connection, Reader, Writer};
 
+/// This trait is for any object that can export keying material.
+///
+/// The terminology comes from [RFC5705](https://datatracker.ietf.org/doc/html/rfc5705)
+/// but doesn't really involve "exporting" key material (in the usual meaning of "export"
+/// -- of moving an artifact from one domain to another) but is best thought of as key
+/// diversification using an existing secret.  That secret is implicit in this interface,
+/// so is assumed to be held by `self`.
+///
+/// There are several such internal implementations, depending on the context
+/// and protocol version.
+pub(crate) trait Exporter: Send + Sync {
+    /// Fills in `output` with derived keying material.
+    ///
+    /// This is deterministic depending on a base secret (implicit in `self`),
+    /// plus the `label` and `context` values.
+    ///
+    /// Must fill in `output` entirely, or return an error.
+    fn derive(&self, output: &mut [u8], label: &[u8], context: Option<&[u8]>) -> Result<(), Error>;
```

Comment:
I think this follows from `export_keying_material()` but it seems odd to have `output` as the first arg instead of the last. We're inconsistent on this in other places (e.g. `crypto/tls13.rs` puts it last in `expand_*()` but places like `crypto/tls12.rs`'s `prf()`, and the pre-existing `export_keying_materia()` put it first.

I think I prefer seeing it last if all things were equal. WDYT?